infra/Pos/Slotting/Types.hs

```diff
@@ -22,15 +38,123 @@ data EpochSlottingData = EpochSlottingData
+createSlottingDataUnsafe :: Map EpochIndex EpochSlottingData -> SlottingData
+createSlottingDataUnsafe epochSlottingDataMap =
+    if M.size epochSlottingDataMap < 2
+        then criticalError
+        else SlottingData epochSlottingDataMap
+  where
+    criticalError = error "It's impossible to create slotting data without two epochs."
+
+-- | Restricted constructor function for the (initial) creation of @SlottingData@.
+createInitSlottingData
+    :: CurrentEpochSlottingData
+    -> NextEpochSlottingData
+    -> SlottingData
+createInitSlottingData psd esd = SlottingData validInitialSlottingData
+  where
+
+    validInitialSlottingData :: Map EpochIndex EpochSlottingData
+    validInitialSlottingData = M.union currentEpochSlottingData nextEpochSlottingData
+
+    currentEpochSlottingData :: Map EpochIndex CurrentEpochSlottingData
+    currentEpochSlottingData = M.singleton 0 psd
+
+    nextEpochSlottingData :: Map EpochIndex NextEpochSlottingData
+    nextEpochSlottingData = M.singleton 1 esd
+
+-- | Get all epoch index.
+getAllEpochIndices :: SlottingData -> [EpochIndex]
+getAllEpochIndices = M.keys . getSlottingDataMap
+
+-- | Get the next epoch index.
+getNextEpochIndex :: SlottingData -> EpochIndex
+getNextEpochIndex = fst . M.findMax . getSlottingDataMap
+
+-- | Get the next epoch slotting data.
+getNextEpochSlottingData :: SlottingData -> EpochSlottingData
+getNextEpochSlottingData = snd . M.findMax . getSlottingDataMap
+
+-- | Get the current epoch index. Next epoch - 1.
+getCurrentEpochIndex :: SlottingData -> EpochIndex
+getCurrentEpochIndex = decreaseEpochIndex . getNextEpochIndex
+  where
+    -- Left for readability.
+    decreaseEpochIndex :: EpochIndex -> EpochIndex
+    decreaseEpochIndex = pred
+
+-- | Get the current epoch slotting data. Next epoch - 1.
+getCurrentEpochSlottingData :: SlottingData -> EpochSlottingData
+getCurrentEpochSlottingData sdp@(getSlottingDataMap -> sd) = sd M.! currentEpochIndex
+  where
+    currentEpochIndex = getCurrentEpochIndex sdp
+
+-- | Lookup the slotting data for an arbitrary `EpochIndex`.
+lookupEpochSlottingData :: EpochIndex -> SlottingData -> Maybe EpochSlottingData
+lookupEpochSlottingData epochIndex slottingData = M.lookup epochIndex slottingData'
+  where
+    slottingData' :: Map EpochIndex EpochSlottingData
+    slottingData' = getSlottingDataMap slottingData
+
+-- | Insert `EpochSlottingData`.
+addEpochSlottingData :: EpochIndex -> EpochSlottingData -> SlottingData -> SlottingData
+addEpochSlottingData epochIndex epochSlottingData slottingData =
+    SlottingData $ M.insert epochIndex epochSlottingData slottingData'
+  where
+    slottingData' :: Map EpochIndex EpochSlottingData
+    slottingData' = getSlottingDataMap slottingData
+
+-- | Compute when the slot started. We give it @SlotId@, the @SlotId@
+-- @EpochSlottingData@ and find when did that @SlotId@ occur.
+computeSlotStart :: Timestamp -> SlotId -> EpochSlottingData -> Timestamp
```

Comment:
Ah, now I understand what you meant. Sure, that does make more sense, I'll fix it.